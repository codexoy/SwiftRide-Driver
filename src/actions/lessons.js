import moment from "moment"

export function getStartAndEndOfDay(date) {
	let timestamp
	let dateString
	if (date.hasOwnProperty("timestamp")) {
		timestamp = date.timestamp
		dateString = date.dateString
	} else {
		timestamp = date.getTime()
		dateString = date.toJSON().slice(0, 10)
	}
	const startOfDay = moment
		.unix(timestamp / 1000) // division by 1000 to get epoch https://stackoverflow.com/questions/3367415/get-epoch-for-a-specific-date-using-javascript
		.utc()
		.startOf("day")
	const endOfDay = moment
		.unix(timestamp / 1000)
		.utc()
		.endOf("day")

	return { startOfDay, endOfDay, dateString }
}
