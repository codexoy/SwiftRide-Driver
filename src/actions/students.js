const _constructAPIUrl = (page, search, orderByColumn, orderByMethod) => {
	let extra = ""
	if (orderByColumn) {
		extra += `&order_by=${orderByColumn} ${orderByMethod}`
	}
	if (search) extra += `&name=${search}`
	return "/teacher/students?limit=20&is_approved=true&page=" + page + extra
}

export const getStudents = async (fetchService, state) => {
	const { page, search, orderByColumn, orderByMethod } = state
	resp = await fetchService.fetch(
		_constructAPIUrl(page, search, orderByColumn, orderByMethod),
		{
			method: "GET"
		}
	)

	return {
		students: resp.json["data"],
		nextUrl: resp.json["next_url"]
	}
}
