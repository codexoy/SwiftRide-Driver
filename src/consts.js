export const TOKEN_KEY = "login_token"
export const REFRESH_TOKEN_KEY = "refresh_token"
export const ROOT_URL = "https://dryvo.herokuapp.com"
export const STORAGE_PREFIX = "Dryvo_"
