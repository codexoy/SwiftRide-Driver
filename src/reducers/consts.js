export const LOAD_FETCH_SERVICE = "LOAD_FETCH_SERVICE"
export const LOGIN = "LOGIN"
export const LOGOUT = "LOGOUT"
