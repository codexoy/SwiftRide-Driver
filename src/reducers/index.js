import { combineReducers } from "redux"
import userReducer from "./user"
import fetchServiceReducer from "./fetch"
import errorsReducer from "./errors"

const appReducer = combineReducers({
	user: userReducer,
	fetchService: fetchServiceReducer,
	errors: errorsReducers
})
export default appReducer
