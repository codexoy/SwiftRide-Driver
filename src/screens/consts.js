import React from "react"
import { Icon } from "react-native-elements"
const options = {
	activeTintColor: "rgb(12,116,244)",
	inactiveTintColor: "black",
	showLabel: false
}

export const tabBarOptions = {
	defaultNavigationOptions: ({ navigation }) => ({
		tabBarIcon: ({ focused, horizontal, tintColor }) => {
			const { routeName } = navigation.state
			let iconName
			let iconType = "material"
			if (routeName === "Home") {
				iconName = "home"
			} else if (routeName === "Notifications") {
				iconName = "notifications"
			} else if (routeName === "Lesson") {
				iconName = "add-circle"
			} else if (routeName === "Students") {
				iconName = "people"
			} else if (routeName == "Schedule") {
				iconName = "date-range"
			} else if (routeName == "Profile") {
				iconName = "person"
			}

			return (
				<Icon
					name={iconName}
					size={25}
					color={tintColor}
					type={iconType}
				/>
			)
		}
	}),
	tabBarOptions: options
}
