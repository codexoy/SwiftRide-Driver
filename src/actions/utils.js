import { Platform, Linking } from "react-native"
import { LOAD_FETCH_SERVICE, API_ERROR, POP_ERROR } from "../reducers/consts"
import { getLatestError } from "../error_handling"
import moment from "moment"
import Storage from "../services/Storage"

export const fetchOrError = (endpoint, params, dispatchError = true) => {
	return async (dispatch, getState) => {
		const { fetchService } = getState()
		try {
			const resp = await fetchService.fetch(endpoint, params)
			return resp
		} catch (error) {
			if (dispatchError) {
				let msg = error || ""
				if (error && error.hasOwnProperty("message"))
					msg = error.message
				dispatch({ type: API_ERROR, error: msg })
			}
			return null
		}
	}
}

export const popLatestError = type => {
	return (dispatch, getState) => {
		const { errors } = getState()
		const error = getLatestError(errors[type])
		if (error) {
			dispatch({ type: POP_ERROR, errorType: type })
			return error
		}
		return null
	}
}

export const loadFetchService = () => {
	return {
		type: LOAD_FETCH_SERVICE
	}
}

export const deepLinkingListener = async func => {
	if (Platform.OS === "android") {
		/// This is called on startup
		let url = await Linking.getInitialURL()
		func({ url: url })
	} else {
		Linking.addEventListener("url", func)
	}
}

export const deepLinkingRemoveListener = async func => {
	Linking.removeEventListener("url", func)
}

export const getHoursDiff = (date, duration) => {
	const start = moment.utc(date).format("HH:mm")
	const end = moment
		.utc(date)
		.add(duration, "minutes")
		.format("HH:mm")

	return { start, end }
}

export const registerDeviceToken = token => {
	return async (dispatch, getState) => {
		const { user } = getState()
		const existing_token = await Storage.getItem(
			"firebase_token_user_" + user.id,
			true
		)
		// we already registered the firebase token.
		// let's check it's expiry and only if it's expired,
		// register again
		if (existing_token && JSON.parse(existing_token).expiry >= new Date())
			return
		return await dispatch(_registerDeviceToken(token))
	}
}

const _registerDeviceToken = fcmToken => {
	return async (dispatch, getState) => {
		if (fcmToken) {
			const { fetchService, user } = getState()
			try {
				const resp = await fetchService.fetch(
					"/user/register_firebase_token",
					{
						method: "POST",
						body: JSON.stringify({
							token: fcmToken
						})
					}
				)
				if (resp.status == 200) {
					let date = new Date()
					const expiry = date.setDate(date.getDate() + 7) // 7 days from now
					await Storage.setItem(
						"firebase_token_user_" + user.id,
						JSON.stringify({ token: fcmToken, expiry }),
						true
					)
				}
			} catch (err) {
				console.log(err)
			}
		}
	}
}
