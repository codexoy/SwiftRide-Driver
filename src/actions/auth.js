import {
	ROOT_URL,
	TOKEN_KEY,
	REFRESH_TOKEN_KEY,
	DEFAULT_ERROR,
	signUpRoles
} from "../consts"
import { Linking } from "react-native"
import Storage from "../services/Storage"
import { LOGIN, LOGOUT, API_ERROR } from "../reducers/consts"
import { fetchOrError } from "./utils"

const loginOrRegister = async (endpoint, params, dispatch, callback) => {
	const resp = await dispatch(fetchOrError(endpoint, params))
	if (resp) {
		await setTokens(resp.json.auth_token, resp.json.refresh_token)
		await dispatch(setUser(resp.json.user))
		await callback(resp.json.user)
	} else await callback(undefined)
}

export const directLogin = (email, password, callback) => {
	return async dispatch => {
		const requestParams = {
			method: "POST",
			body: JSON.stringify({ email, password })
		}
		await loginOrRegister(
			"/login/direct",
			requestParams,
			dispatch,
			callback
		)
	}
}

export const register = (params, callback, role = "") => {
	return async dispatch => {
		var data = new FormData()
		Object.keys(params).forEach(key => data.append(key, params[key]))
		const requestParams = {
			method: "POST",
			headers: {
				"Content-Type": "multipart/form-data"
			},
			body: data
		}
		const newCallback = async user => {
			if (user) {
				let roleRequest
				if (role == signUpRoles.student) {
					roleRequest = await dispatch(
						fetchOrError(
							"/user/make_student?teacher_id=" +
								params.teacher_id,
							{
								method: "GET"
							}
						)
					)
				} else if (role == signUpRoles.teacher) {
					roleRequest = await dispatch(
						fetchOrError("/user/make_teacher?", {
							method: "POST",
							body: JSON.stringify({
								price: params.price,
								lesson_duration: params.duration
							})
						})
					)
				}
				if (roleRequest) {
					await callback(user)
					return
				}
			}
			await callback(undefined)
		}

		await loginOrRegister(
			"/login/register",
			requestParams,
			dispatch,
			newCallback
		)
	}
}

const setTokens = async (token, refresh_token) => {
	await Storage.setItem(TOKEN_KEY, token, true)
	await Storage.setItem(REFRESH_TOKEN_KEY, refresh_token, true)
}

export const setUser = user => {
	return dispatch => {
		dispatch({ type: LOGIN, user: user })
	}
}

export const logout = (callback = async () => {}) => {
	return async dispatch => {
		console.log("Logging out")
		await Storage.removeItem(TOKEN_KEY, true)
		await Storage.removeItem(REFRESH_TOKEN_KEY, true)
		await callback()
		setTimeout(() => {
			dispatch({ type: LOGOUT })
		}, 100)
	}
}

export const fetchUser = (callback = () => {}) => {
	/* effectively checks if the token is valid */
	return async dispatch => {
		const resp = await dispatch(
			fetchOrError("/user/me", { method: "GET" }, false)
		)
		if (resp) {
			dispatch(setUser(resp.json.user))
			await callback(resp.json.user)
		} else await callback(undefined)
	}
}

export const exchangeToken = (token, callback) => {
	return async dispatch => {
		try {
			const resp = await dispatch(
				fetchOrError("/login/exchange_token", {
					method: "POST",
					body: JSON.stringify({
						exchange_token: token
					})
				})
			)
			await setTokens(resp.json.auth_token, resp.json.refresh_token)
			await dispatch(fetchUser(callback))
		} catch (error) {
			dispatch({ type: API_ERROR, error: DEFAULT_ERROR })
		}
	}
}

export const openFacebook = (token = "") => {
	let url = ROOT_URL + "/login/facebook"
	if (token) {
		url += "?token=" + token
	}
	Linking.openURL(url)
}
