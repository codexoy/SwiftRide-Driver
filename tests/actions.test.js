import * as utils from "../src/actions/utils"
import * as auth from "../src/actions/auth"
import * as students from "../src/actions/students"
import * as lessons from "../src/actions/lessons"
import FetchService from "../src/services/Fetch"
import Storage from "../src/services/Storage"
import { Platform } from "react-native"
import mockStore from "redux-mock-store"
import { TOKEN_KEY, REFRESH_TOKEN_KEY, DEFAULT_IMAGE } from "../src/consts"
import { API_ERROR } from "../src/reducers/consts"

let user, store
beforeAll(() => {
	user = { id: 1, name: "test" }
})
beforeEach(() => {
	store = mockStore({ fetchService: new FetchService(), user })
})
describe("utils.js", () => {
	it("should register device token", async () => {
		const mockToken = "myMockToken"
		const response = {}
		fetch.mockResponseSuccess(JSON.stringify(response))
		let registerToken = await store.dispatch(
			utils.registerDeviceToken(mockToken)
		)
		let token = await Storage.getItem(
			"firebase_token_user_" + user.id,
			true
		)
		token = JSON.parse(token)
		expect(token.token).toEqual("myMockToken")
		store.clearActions()

		// now its supposed to return null
		registerToken = await store.dispatch(
			utils.registerDeviceToken(mockToken)
		)
		expect(registerToken).toBe(undefined)
		store.clearActions()
	})
	it("should override the token", async () => {
		// now let's change the expiry and see that it overrides our token
		let d = new Date()
		d.setDate(d.getDate() - 1)
		await Storage.setItem(
			"firebase_token_user_" + user.id,
			JSON.stringify({ token: null, d }),
			true
		)
		fetch.mockResponseSuccess(JSON.stringify({}))
		const registerToken = await store.dispatch(
			utils.registerDeviceToken("newToken")
		)
		let token = await Storage.getItem(
			"firebase_token_user_" + user.id,
			true
		)
		token = JSON.parse(token)
		expect(token.token).toEqual("newToken")
		store.clearActions()
	})

	it("should dispatch pop error", () => {
		const storeWithErrors = mockStore({
			errors: { [API_ERROR]: ["test", "test2"] }
		})
		const error = storeWithErrors.dispatch(utils.popLatestError(API_ERROR))
		expect(error).toEqual("test2")
		expect(storeWithErrors.getActions()).toMatchSnapshot()
	})

	it("should return response from request", async () => {
		const example = { test: "test" }
		fetch.mockResponseSuccess(JSON.stringify(example))
		const resp = await store.dispatch(utils.fetchOrError("/test", {}))
		expect(resp.json).toEqual(example)
		store.clearActions()
	})

	it("should dispatch API ERROR", async () => {
		fetch.mockResponseFailure()
		await store.dispatch(utils.fetchOrError("/test", {}))
		expect(store.getActions()).toMatchSnapshot()
		store.clearActions()
	})

	it("should not dispatch API ERROR", async () => {
		fetch.mockResponseFailure()
		await store.dispatch(utils.fetchOrError("/test", {}, false))
		expect(store.getActions()).toMatchSnapshot()
		store.clearActions()
	})

	it("should return load fetch service", () => {
		expect(utils.loadFetchService()).toMatchSnapshot()
	})

	it("should set linking listener", async () => {
		const func = jest.fn()
		Platform.OS = "android"
		await utils.deepLinkingListener(func)
		expect(func).toBeCalled()
	})

	it("should return default image", () => {
		let image = utils.getUserImage({ name: "test" })
		expect(image).toEqual(DEFAULT_IMAGE)
		image = utils.getUserImage(null)
		expect(image).toEqual(DEFAULT_IMAGE)
	})

	it("should return user's image", () => {
		const image = utils.getUserImage({ image: "test" })
		expect(image).toEqual("test")
	})

	it("should dispatch API ERROR when trying to upload image", async () => {
		fetch.mockResponseFailure()
		await store.dispatch(utils.uploadUserImage({ uri: "test" }))
		expect(store.getActions()).toMatchSnapshot()
		store.clearActions()
	})

	it("should dispatch CHANGE_USER_IMAGE", async () => {
		const example = { image: "test" }
		fetch.mockResponseSuccess(JSON.stringify(example))
		await store.dispatch(utils.uploadUserImage({ uri: "test" }))
		expect(store.getActions()).toMatchSnapshot()
		store.clearActions()
	})
})

describe("auth.js", () => {
	describe("fetchUser test", () => {
		const callback = jest.fn(user => {
			console.log(user)
		})
		it("should (dispatch LOGIN & call callback with user)", async () => {
			fetch.mockResponseSuccess(JSON.stringify({ user }))
			await store.dispatch(auth.fetchUser(callback))
			expect(callback).toHaveBeenCalledWith(
				expect.objectContaining({ name: user.name })
			)
			expect(store.getActions()).toMatchSnapshot()
		})
		it("should call callback with undefined", async () => {
			fetch.mockResponseFailure()
			await store.dispatch(auth.fetchUser(callback))
			expect(callback).toHaveBeenLastCalledWith(expect.undefined)
		})
	})

	describe("exchangeToken tests", () => {
		const callback = jest.fn(user => {
			console.log(user)
		})
		it("should dispatch error", async () => {
			fetch.mockResponseFailure()
			await store.dispatch(auth.exchangeToken("exchange-token", callback))
			expect(store.getActions()).toMatchSnapshot()
			store.clearActions()
		})
		it("should set tokens and dispatch login user", async () => {
			const response = {
				auth_token: "auth-test",
				refresh_token: "refresh-test"
			}
			fetch.mockResponseSuccess(JSON.stringify(response))
			fetch.mockResponseSuccess(JSON.stringify({ user }))
			await store.dispatch(auth.exchangeToken("exchange-token", callback))
			const token = await Storage.getItem(TOKEN_KEY, true)
			const refresh = await Storage.getItem(REFRESH_TOKEN_KEY, true)
			expect(token).toEqual("auth-test")
			expect(refresh).toEqual("refresh-test")
			expect(store.getActions()).toMatchSnapshot()
		})
	})

	describe("logout tests", () => {
		it("should dispatch logout and remove tokens", async () => {
			const callback = jest.fn(() => {})
			let ret = await store.dispatch(auth.logout(callback))
			expect(callback).toBeCalled()
			const token = await Storage.getItem(TOKEN_KEY, true)
			const refresh = await Storage.getItem(REFRESH_TOKEN_KEY, true)
			expect(token).toEqual(null)
			expect(refresh).toEqual(null)
			expect(store.getActions()).toMatchSnapshot()
		})
	})

	describe("register tests", () => {
		const callback = jest.fn()
		const response = {
			user: { id: 1, name: "test" },
			auth_token: "auth-test",
			refresh_token: "refresh-test"
		}
		it("should set tokens, dispatch login user and call callback", async () => {
			fetch.mockResponseSuccess(JSON.stringify(response))
			let ret = await store.dispatch(
				auth.register(
					{
						email: "t@t.com",
						name: "t",
						area: "t",
						password: "test",
						image: "test"
					},
					callback
				)
			)
			expect(callback).toHaveBeenCalled()
			const token = await Storage.getItem(TOKEN_KEY, true)
			const refresh = await Storage.getItem(REFRESH_TOKEN_KEY, true)
			expect(token).toEqual("auth-test")
			expect(refresh).toEqual("refresh-test")
			expect(store.getActions()).toMatchSnapshot()
			store.clearActions()
		})

		it("should dispatch default error", async () => {
			fetch.mockResponseFailure()
			let ret = await store.dispatch(auth.register({}, callback))
			expect(store.getActions()).toMatchSnapshot()
		})
	})

	describe("directLogin tests", () => {
		const callback = jest.fn()
		const response = {
			user: { id: 1, name: "test" },
			auth_token: "auth-test",
			refresh_token: "refresh-test"
		}
		it("should set tokens, dispatch login user and call callback", async () => {
			fetch.mockResponseSuccess(JSON.stringify(response))
			let ret = await store.dispatch(
				auth.directLogin("t@t.com", "test", callback)
			)
			expect(callback).toHaveBeenCalled()
			const token = await Storage.getItem(TOKEN_KEY, true)
			const refresh = await Storage.getItem(REFRESH_TOKEN_KEY, true)
			expect(token).toEqual("auth-test")
			expect(refresh).toEqual("refresh-test")
			expect(store.getActions()).toMatchSnapshot()
			store.clearActions()
		})

		it("should dispatch default error", async () => {
			fetch.mockResponseFailure()
			let ret = await store.dispatch(
				auth.directLogin("test", "test", callback)
			)
			expect(store.getActions()).toMatchSnapshot()
		})
	})
})

describe("students.js", () => {
	describe("getStudents", () => {
		it("should return dict with students and nextUrl", async () => {
			const data = [
				{ id: 1, date: new Date() },
				{ id: 2, date: new Date() }
			]
			fetch.mockResponseSuccess(
				JSON.stringify({
					data
				})
			)
			const resp = await students.getStudents(new FetchService(), {
				page: 1,
				search: "",
				orderByColumn: "",
				orderByMethod: "asc"
			})
			expect(resp["students"][0]["id"]).toEqual(1)
		})
	})
})

describe("lessons.js", () => {
	describe("getDateAndString", () => {
		it("should return dict with 2 values", async () => {
			const ret = lessons.getDateAndString(new Date())
			expect(Object.keys(ret).length).toEqual(2)
		})
	})

	describe("getPayments", () => {
		it("should return dict with 2 values and make a fetch request", async () => {
			const response = { data: [{ amount: 30 }, { amount: 100 }] }
			fetch.mockResponseSuccess(JSON.stringify(response))
			const ret1 = await lessons.getPayments(new FetchService())
			expect(ret1.payments).toEqual(response.data)
			expect(Object.keys(ret1).length).toEqual(2)
		})
	})
})
