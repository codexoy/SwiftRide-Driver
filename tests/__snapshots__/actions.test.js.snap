// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth.js directLogin tests should dispatch default error 1`] = `
Array [
  Object {
    "error": "",
    "type": "API_ERROR",
  },
]
`;

exports[`auth.js directLogin tests should set tokens, dispatch login user and call callback 1`] = `
Array [
  Object {
    "type": "LOGIN",
    "user": Object {
      "id": 1,
      "name": "test",
    },
  },
]
`;

exports[`auth.js exchangeToken tests should dispatch error 1`] = `
Array [
  Object {
    "error": "תקלה בהתחברות, אנא נסו שנית.",
    "type": "API_ERROR",
  },
]
`;

exports[`auth.js exchangeToken tests should set tokens and dispatch login user 1`] = `
Array [
  Object {
    "type": "LOGIN",
    "user": Object {
      "id": 1,
      "name": "test",
    },
  },
]
`;

exports[`auth.js fetchUser test should (dispatch LOGIN & call callback with user) 1`] = `
Array [
  Object {
    "type": "LOGIN",
    "user": Object {
      "id": 1,
      "name": "test",
    },
  },
]
`;

exports[`auth.js logout tests should dispatch logout and remove tokens 1`] = `
Array [
  Object {
    "type": "LOGOUT",
  },
]
`;

exports[`auth.js register tests should dispatch default error 1`] = `
Array [
  Object {
    "error": "",
    "type": "API_ERROR",
  },
]
`;

exports[`auth.js register tests should set tokens, dispatch login user and call callback 1`] = `
Array [
  Object {
    "type": "LOGIN",
    "user": Object {
      "id": 1,
      "name": "test",
    },
  },
]
`;

exports[`utils.js should return load fetch service 1`] = `
Object {
  "type": "LOAD_FETCH_SERVICE",
}
`;
